pkg/ruler/api.go

```diff
@@ -268,3 +282,274 @@ func (r *Ruler) alerts(w http.ResponseWriter, req *http.Request) {
+				ruleName = r.Alert
+			} else {
+				ruleName = r.Record
+			}
+			errs = append(errs, &rulefmt.Error{
+				Group:    g.Name,
+				Rule:     i,
+				RuleName: ruleName,
+				Err:      err,
+			})
+		}
+	}
+
+	return errs
+}
+
+// parseNamespace parses the namespace from the provided set of params, in this
+// api these params are derived from the url path
+func parseNamespace(params map[string]string) (string, error) {
+	namespace, exists := params["namespace"]
+	if !exists {
+		return "", ErrNoNamespace
+	}
+
+	namespace, err := url.PathUnescape(namespace) // namespaces needs to be unescaped if in the URL
+	if err != nil {
+		return "", err
+	}
+
+	return namespace, nil
+}
+
+// parseGroupName parses the group name from the provided set of params, in this
+// api these params are derived from the url path
+func parseGroupName(params map[string]string) (string, error) {
+	groupName, exists := params["groupName"]
+	if !exists {
+		return "", ErrNoGroupName
+	}
+
+	groupName, err := url.PathUnescape(groupName) // groupName needs to be unescaped if in the URL
+	if err != nil {
+		return "", err
+	}
+
+	return groupName, nil
+}
+
+func (r *Ruler) listRules(w http.ResponseWriter, req *http.Request) {
+	logger := util.WithContext(req.Context(), util.Logger)
+	userID, err := user.ExtractOrgID(req.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	vars := mux.Vars(req)
+
+	// Parse the namespace from the url path parameters and return
+	// a 400 if it is invalid and return a full set of rules if no
+	// namespace is provided
+	namespace, err := parseNamespace(vars)
+	if err != nil {
+		// If a namespace is not provided continue as usual and return a full set of rules
+		if err != ErrNoNamespace {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+		level.Debug(logger).Log("msg", "retrieving rule groups with namespace", "userID", userID, "namespace", namespace)
+	}
+
+	level.Debug(logger).Log("msg", "retrieving rule groups from rule store", "userID", userID)
+	rgs, err := r.store.ListRuleGroups(req.Context(), userID, namespace)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	level.Debug(logger).Log("msg", "retrieved rule groups from rule store", "userID", userID, "num_namespaces", len(rgs))
+
+	if len(rgs) == 0 {
+		level.Info(logger).Log("msg", "no rule groups found", "userID", userID)
+		http.Error(w, ErrNoRuleGroups.Error(), http.StatusNotFound)
+		return
+	}
+
+	formatted := rgs.Formatted()
+
+	d, err := yaml.Marshal(&formatted)
+	if err != nil {
+		level.Error(logger).Log("msg", "error marshalling yaml rule groups", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/yaml")
+	if _, err := w.Write(d); err != nil {
+		level.Error(logger).Log("msg", "error writing yaml response", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```

Comment:
If write failed, there is no need to try sending another status code and error message to the client.

Reply:
👍 makes sense